lib/rust/ensogl/core/src/system/gpu/context/extension.rs

```diff
@@ -0,0 +1,81 @@
+//! WebGL extensions management.
+
+use crate::prelude::*;
+
+use crate::system::gpu::data::GlEnum;
+
+use web_sys::WebGl2RenderingContext;
+use web_sys::WebGlProgram;
+
+
+
+// ==================
+// === Extensions ===
+// ==================
+
+/// Set of all extensions that we try to enable after acquiring the context.
+#[derive(Debug, Clone, Deref)]
+pub struct Extensions {
+    rc: Rc<ExtensionsData>,
+}
+
+impl Extensions {
+    /// Constructor.
+    pub fn init(context: &WebGl2RenderingContext) -> Self {
+        Self { rc: Rc::new(ExtensionsData::init(context)) }
+    }
+}
+
+/// Internal representation of [`Extensions`].
+#[derive(Copy, Clone, Debug)]
+#[allow(missing_docs)]
+pub struct ExtensionsData {
+    pub khr_parallel_shader_compile: Option<KhrParallelShaderCompile>,
+}
+
+impl ExtensionsData {
+    /// Constructor.
+    fn init(context: &WebGl2RenderingContext) -> Self {
+        let khr_parallel_shader_compile = KhrParallelShaderCompile::init(context);
+        Self { khr_parallel_shader_compile }
+    }
+}
+
+
+
+// ================================
+// === KhrParallelShaderCompile ===
+// ================================
+
+/// The `KHR_parallel_shader_compile` extension is used to poll shader compilation status without
+/// blocking. To learn more, see:
+/// [https://www.khronos.org/registry/webgl/extensions/KHR_parallel_shader_compile]
+#[derive(Clone, Copy, Debug)]
+#[allow(missing_docs)]
+pub struct KhrParallelShaderCompile {
+    pub completion_status_khr: GlEnum,
+}
+
+impl KhrParallelShaderCompile {
+    /// Try to obtain the extension.
+    pub fn init(context: &WebGl2RenderingContext) -> Option<Self> {
+        let ext = context.get_extension("KHR_parallel_shader_compile").ok()??;
+        let completion_status_khr = GlEnum(
+            js_sys::Reflect::get(&ext, &"COMPLETION_STATUS_KHR".into()).ok()?.as_f64()? as u32,
+        );
+        Some(Self { completion_status_khr })
+    }
+
+    /// Asynchronously check if the job is ready.
+    pub fn is_ready(&self, context: &WebGl2RenderingContext, program: &WebGlProgram) -> bool {
+        let param = self.completion_status_khr;
+        context.get_program_parameter(program, *param).as_bool().unwrap_or_else(|| {
+            REPORTABLE_WARNING!(
+                "context.getProgramParameter returned non bool value for KHR \
+                Parallel Shader Compile status check. This should never happen, however, it should \
+                not cause visual artifacts. Reverting to non-parallel mode."
```

Comment:
The ` Reverting to non-parallel mode.` is not actually implemented in this method, though. So, I'm wondering whether this info should be emitted where it is actually done to avoid future refactoring breaking this warning message.

Reply:
This is a very good comment. Moved this thing to the right place. Thanks for spotting it.